lib/codecs/src/decoding/format/protobuf.rs

```diff
@@ -0,0 +1,311 @@
+pub struct ProtobufDeserializerConfig {
+    /// Path to desc file
+    desc_file: String,
+
+    /// message type. e.g package.message
+    message_type: String,
+}
+
+impl ProtobufDeserializerConfig {
+    /// Build the `ProtobufDeserializer` from this configuration.
+    pub fn build(&self) -> ProtobufDeserializer {
+        Into::<ProtobufDeserializer>::into(self)
+    }
+
+    /// Return the type of event build by this deserializer.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema produced by the deserializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        match log_namespace {
+            LogNamespace::Legacy => {
+                let mut definition =
+                    schema::Definition::empty_legacy_namespace().unknown_fields(Kind::json()); // TODO: create kind::protobuf?
+
+                if let Some(timestamp_key) = log_schema().timestamp_key() {
+                    definition = definition.try_with_field(
+                        timestamp_key,
+                        // The PROTOBUF decoder will try to insert a new `timestamp`-type value into the
+                        // "timestamp_key" field, but only if that field doesn't already exist.
+                        Kind::json().or_timestamp(), // TODO: create kind::protobuf?
+                        Some("timestamp"),
+                    );
+                }
+                definition
+            }
+            LogNamespace::Vector => {
+                schema::Definition::new_with_default_metadata(Kind::json(), [log_namespace])
+                // TODO: create kind::protobuf?
+            }
+        }
+    }
+}
+
+impl ProtobufDeserializerConfig {
+    /// Creates a new `ProtobufDeserializerConfig`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+/// Deserializer that builds `Event`s from a byte frame containing PROTOBUF.
+#[derive(Debug, Clone)]
+pub struct ProtobufDeserializer {
+    message_descriptor: MessageDescriptor,
+}
+
+impl ProtobufDeserializer {
+    /// Creates a new `ProtobufDeserializer`.
+    pub fn new(desc_file: String, message_type: String) -> Self {
+        // TODO: handle 'expect' in a better way
+        let b = fs::read(desc_file.clone())
+            .expect(&format!("Failed to open protobuf desc file '{desc_file}'"));
+        let pool = DescriptorPool::decode(b.as_slice())
+            .expect(&format!("Failed to parse protobuf desc file '{desc_file}'"));
+        Self {
+            message_descriptor: pool.get_message_by_name(&message_type).expect(&format!(
+                "The message type '{message_type}' could not be found in '{desc_file}'"
+            )),
+        }
+    }
+}
+
+impl Deserializer for ProtobufDeserializer {
+    fn parse(
+        &self,
+        bytes: Bytes,
+        log_namespace: LogNamespace,
+    ) -> vector_common::Result<SmallVec<[Event; 1]>> {
+        if bytes.is_empty() {
```

Comment:
Is it OK to get empty `bytes` here?

Reply:
thanks for noticing, I did a small test in a demo code, and `protobuf` does support empty buffer as normal message, I fixed the code and the test.